mantidproject/mantid
# Refactor and add ALT format to SaveGSS

Description of work.
As part of discussions with ENGINX we have been asked to implement the GSAS ALT format. This allows them to continue using existing Python scripts for analysis of textured banks whilst migrating over to using Mantid.

This PR:

- Updates the unit tests for `SaveGSS` to use the `FileComparisonHelper` - that way we check the entire file instead of just looking at snippets (as GSAS can be white space sensitive)
- Adds the option to select ALT format in addition to FXYE when using RALF headers

-----

- Refactors the existing code :- 
- Move the implementation out of `exec` into separate functions 
- Additional sanity checks
- Makes several parts of the alg parallel such as checking each spectra has a valid detector....etc.
- The progress bar actually shows progress through the algorithm now and doesn't get "stuck" at 100%

**To test:**

On both master / release and this branch:

- Open the following focused data: [POLARIS98532.zip](https://github.com/mantidproject/mantid/files/1176373/POLARIS98532.zip)
- Run `SaveGSS` on the:
- Workspace group
- Individual workspace
- With various options such as `Append` set to on / off
- Compare the file output from both. (On Linux Meld works really well. On Windows Notepad++ with the Compare plugin also works)
- Run the following:
```
CreateSampleWorkspace(OutputWorkspace='ws1', NumBanks=5, BankPixelWidth=1)
```
- Save this out with SaveGSS with `SplitFiles` On / Off, it should correctly split into 5 separate files


The file names will differ slightly as we no longer append `-0` to the filename when we are not splitting files. 

Fixes #19413 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/pull/20100/files#diff-d43ebaec89d7bc36364c8bf4c9f97673)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
